### ISSUE TITLE
SAM integration with lightning trainer

### ISSUE THREAD

**Issue by yazdanimehdi** [Lightning-AI/lightning#17650]
### Description & Motivation

In the recent literature, sharpness aware minimization shows promising results in vision neural networks.
Multiple issues prevent us from using this optimization method with the lightning trainer.
1. two-step optimization procedure.
2. gradient accumulation cannot be done like normal in this scenario.

Paper:
SAM: https://arxiv.org/pdf/2010.01412.pdf
GSAM: https://arxiv.org/pdf/2203.08065.pdf

### Pitch

There would be great if we have a flag in the trainer or optimizer function of the lightning module to specify the usage of these algorithms for lightning to handle it.

cc @borda @justusschock @awaelchli

**Comment by carmocca** [comment]
Duplicate of https://github.com/Lightning-AI/lightning/issues/16143
